README.md
# Notice

This cookbook's changes have been merged upstream into [gamegos/supervisor-cookbook](https://github.com/gamegos/supervisor-cookbook). Use that instead.

# buyerquest-supervisor

This cookbook installs and configures [Supervisor](https://github.com/Supervisor/supervisor).

# Requirements

## Platforms

- Ubuntu 14.04+
- Centos 7+
- Debian 7+

## Chef

- Chef 12+

## Cookbooks

- poise-python
- poise-service

# Usage

Add to your `Berksfile`:

```ruby
cookbook 'buyerquest-supervisor', github: 'buyerquest/buyerquest-supervisor'
```

And to `metadata.rb`:
```ruby
depends 'buyerquest-supervisor'
```

Now in your Chef recipe, here's a quick example of installing the supervisord and adding some programs.

```ruby
supervisor_install '3.3'

supervisor_config 'supervisor'

supervisor_process 'cat' do
  type 'program'
  command '/bin/cat'
end

supervisor_process 'vi' do
  type 'program'
  command '/usr/bin/vi'
end

supervisor_group 'cat' do
  programs ['cat']
end

supervisor_service 'supervisor'
```

# Recipes

- **default** - installs the supervisord.

# Resources

### supervisor_install

Install Supervisor with pip.

#### Actions

- `:create`

#### Examples

```ruby
supervisor_install '3.3'
```

### supervisor_config

Define configurations of supervisord.

#### Actions

- `:create`

#### Properties

- `socket_file` - (is: String)
- `unix_http_server_chmod` - (is: String)
- `unix_http_server_chown` - (is: String)
- `unix_http_server_username` - (is: String)
- `unix_http_server_password` - (is: String)
- `supervisord_config_directory` - (is: String)
- `supervisord_log_directory` - (is: String)
- `supervisord_logfile` - (is: String)
- `supervisord_logfile_maxbytes` - (is: String)
- `supervisord_logfile_backups` - (is: Integer)
- `supervisord_loglevel` - (is: String)
- `supervisord_pidfile` - (is: String)
- `supervisord_nodaemon` - (is: Boolean)
- `supervisord_minfds` - (is: Integer)
- `supervisord_minprocs` - (is: Integer)
- `supervisord_nocleanup` - (is: Boolean)
- `supervisord_user` - (is: String)
- `supervisord_umask` - (is: String)
- `supervisord_identifier` - (is: String)
- `supervisord_strip_ansi` - (is: Boolean)
- `supervisord_environment` - (is: Hash)
- `inet_port` - (is: String)
- `inet_username` - (is: String)
- `inet_password` - (is: String)
- `include_files` - (is: [String, Array])

#### Examples

```ruby
# Create a custom config
supervisor_config 'supervisor' do
  supervisord_config_directory '/etc/supervisor'
  socket_file '/run/supervisor.sock'
  inet_port '0.0.0.0:9010'
  inet_username 'randy'
  inet_password 'elite'
  supervisord_environment NODE_CONFIG_PORT: '5000', NODE_CONFIG_KEY: '1DFS123SDFK'
  include_files [ '/etc/supervisor/*.conf', '/etc/supervisor/*.ini' ]
  action :create
end
```

```ruby
# Create a default config
supervisor_config 'supervisor'
```

#### Notes

The name property of the resource should always just be `supervisor`.

### supervisor_process

Creates a process for supervisor. The process may be a 'program', 'eventlistener' or 'fcgi-program'.

#### Actions

- `:create`

#### Properties

- `type` - (is: String)
- `command` - (is: String)
- `process_name` - (is: String)
- `numprocs` - (is: Integer)
- `numprocs_start` - (is: Integer)
- `priority` - (is: Integer)
- `autostart` - (is: Boolean)
- `autorestart` - (is: [String, Boolean, Symbol])
- `startsecs` - (is: Integer)
- `startretries` - (is: Integer)
- `exitcodes` - (is: Array)
- `stopsignal` - (is: [String, Symbol])
- `stopwaitsec` - (is: Integer)
- `stopasgroup` - (is: Boolean)
- `killasgroup` - (is: Boolean)
- `user` - (is: String)
- `redirect_stderr` - (is: Boolean)
- `environment` - (is: Hash)
- `directory` - (is: [String, Nil])
- `umask` - (is: [String, Nil])
- `serverurl` - (is: String)
- `stdout_logfile` - (is: String)
- `stdout_logfile_maxbytes` - (is: String)
- `stdout_logfile_backups` - (is: Integer)
- `stdout_capture_maxbytes` - (is: String)
- `stdout_events_enabled` - (is: Boolean)
- `stderr_logfile` - (is: String)
- `stderr_logfile_maxbytes` - (is: String)
- `stderr_logfile_backups` - (is: Integer)
- `stderr_capture_maxbytes` - (is: String)
- `stderr_events_enabled` - (is: Boolean)
- `eventlistener_buffer_size` - (is: Integer)
- `eventlistener_events` - (is: Array)
- `eventlistener_result_handler` - (is: String)
- `fcgi_socket` - (is: String)
- `fcgi_socket_owner` - (is: [String, Nil])
- `fcgi_socket_mode` - (is: String)

#### Examples

```ruby
supervisor_process 'my-nginx' do
  type 'program'
  command "/usr/sbin/nginx -g 'daemon off;'"
  numprocs 5
  process_name "%(program_name)s_%(process_num)02d"
  autorestart true
  user 'root'
end
```

```ruby
supervisor_process 'cat' do
  type 'program'
  command '/bin/cat'
end
```

### supervisor_group

Creates a group for supervisor.

#### Actions

- `:create`

#### Properties

- `programs` - (is: Array)
- `priority` - (is: Integer)

#### Examples

```ruby
supervisor_group 'my-web-servers' do
  programs [ 'my-nginx', 'my-apache' ]
  priorty 100
end
```

### supervisor_service

Creates Supervisor service for your systems.

#### Actions

- `:create`

#### Examples

```ruby
supervisor_service 'supervisor'
```

#### Note

The name property of the resource should always just be `supervisor`.


CHANGELOG.md
# buyerquest-supervisor CHANGELOG

This file is used to list changes made in each version of the buyerquest-supervisor cookbook.

# 1.0.0

Update

- Add Chef 12 support
- Add tests
  - Testing Ubuntu and CentOS
- Update documentation
- Fixes some bugs


# 0.1.0

Initial release.

- change 0
- change 1


recipes/test.rb
#
# Cookbook:: buyerquest-supervisor
# Recipe:: test
#
# Copyright:: 2018, The Authors, All Rights Reserved.
supervisor_install '3.3'

supervisor_config 'supervisor'

supervisor_process 'cat' do
  type 'program'
  command '/bin/cat'
end

supervisor_process 'vi' do
  type 'program'
  command '/usr/bin/vi'
end

supervisor_group 'cat' do
  programs ['cat']
end

supervisor_group 'vi' do
  programs ['vi']
end

supervisor_service 'supervisor'


spec/unit/recipes/default_spec.rb
#
# Cookbook:: buyerquest-supervisor
# Spec:: default
#
# Copyright:: 2018, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'buyerquest-supervisor::test' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7.4.1708' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.4.1708')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/integration/default/default_test.rb
# Inspec test for recipe buyerquest-supervisor::test

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

if os.family == 'debian'
  supervisor_bin_directory = '/usr/local/bin'
  python_package = 'python2.7'
elsif os.family == 'redhat'
  supervisor_bin_directory = '/usr/bin'
  python_package = 'python'
end

describe package(python_package) do
  it { should be_installed }
end

describe pip('supervisor') do
  it { should be_installed }
end

describe service('supervisor') do
  it { should be_running }
end

describe file("#{supervisor_bin_directory}/supervisord") do
  it { should exist }
end

describe file("#{supervisor_bin_directory}/supervisorctl") do
  it { should exist }
end

describe file('/etc/supervisor/supervisord.conf') do
  it { should exist }
end

describe command('supervisord -h') do
  its(:stdout) { should match 'supervisord -- run a set of applications as daemons.' }
  its('exit_status') { should eq 0 }
end

describe command('supervisorctl -h') do
  its(:stdout) { should match 'supervisorctl -- control applications run by supervisord from the cmd line' }
  its('exit_status') { should eq 0 }
end

%w[cat vi].each do |process|
  describe file("/etc/supervisor/program_#{process}.conf") do
    it { should exist }
  end

  describe file("/etc/supervisor/group_#{process}.conf") do
    it { should exist }
  end

  describe command("supervisorctl status #{process}") do
    its(:stdout) { should match "#{process} *STOPPED\s" }
    its('exit_status') { should eq 0 }
  end

  describe command("supervisorctl start #{process}") do
    its(:stdout) { should match "#{process}: started" }
    its('exit_status') { should eq 0 }
  end

  describe command("supervisorctl status #{process}") do
    its(:stdout) { should match "#{process} *RUNNING\s" }
    its('exit_status') { should eq 0 }
  end

  describe command("supervisorctl stop #{process}") do
    its(:stdout) { should match "#{process}: stopped" }
    its('exit_status') { should eq 0 }
  end
end


metadata.rb
name 'buyerquest-supervisor'
maintainer 'Andrew Bobulsky'
license 'GPL-3.0'
description 'Installs/Configures supervisor'
long_description 'Installs/Configures supervisor'
chef_version '>= 12.0'
%w(amazon centos debian redhat ubuntu).each do |os|
  supports os
end
version '1.0.0'

# The `issues_url` points to the location where issues for this cookbook are
# tracked.  A `View Issues` link will be displayed on this cookbook's page when
# uploaded to a Supermarket.
#
issues_url 'https://github.com/buyerquest/supervisor-cookbook/issues'

# The `source_url` points to the development repository for this cookbook.  A
# `View Source` link will be displayed on this cookbook's page when uploaded to
# a Supermarket.
#
source_url 'https://github.com/buyerquest/supervisor-cookbook'

depends 'poise-python', '~> 1.7.0'
depends 'poise-service', '~> 1.5.2'


resources/group.rb
# To learn more about Custom Resources, see https://docs.chef.io/custom_resources.html
resource_name :supervisor_group

property :name, String, name_attribute: true
property :programs, Array, required: true
property :priority, Integer, default: 999
property :template, String, default: 'buyerquest-supervisor'

action :create do
  declare_resource(:template, "supervisor_group_#{new_resource.name}") do
    cookbook new_resource.template
    path(lazy { "#{node.run_state['supervisor']['directory']}/group_#{new_resource.name}.conf" })
    source 'group.conf.erb'
    owner 'root'
    group 'root'
    mode '644'
    variables group: new_resource
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end

action :delete do
  file 'delete specific group configuration file' do
    path(lazy { "#{node.run_state['supervisor']['directory']}/group_#{new_resource.name}.conf" })
    action :delete
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end


resources/config.rb
# To learn more about Custom Resources, see https://docs.chef.io/custom_resources.html
resource_name :supervisor_config

property :name, String, name_property: true
property :supervisord_config_directory, String, default: '/etc/supervisor'
property :socket_file, String, default: '/var/run/supervisor.sock'

property :unix_http_server_chmod, String, default: '700'
property :unix_http_server_chown, String, default: 'root:root'
property :unix_http_server_username, [String, NilClass], default: nil
property :unix_http_server_password, [String, NilClass], default: nil

property :supervisord_log_directory, String, default: '/var/log/supervisor'
property :supervisord_logfile, String, default: lazy {
  "#{supervisord_log_directory}/supervisord.log"
}
property :supervisord_logfile_maxbytes, String, default: '50MB'
property :supervisord_logfile_backups, Integer, default: 10
property :supervisord_loglevel, String, default: 'info'
property :supervisord_pidfile, String, default: '/var/run/supervisord.pid'
property :supervisord_nodaemon, [TrueClass, FalseClass], default: false
property :supervisord_minfds, Integer, default: 1024
property :supervisord_minprocs, Integer, default: 200
property :supervisord_nocleanup, [TrueClass, FalseClass], default: false
property :supervisord_user, [String, NilClass], default: nil
property :supervisord_umask, [String, NilClass], default: nil
property :supervisord_identifier, String, default: 'supervisor'
property :supervisord_strip_ansi, [TrueClass, FalseClass], default: false
property :supervisord_environment, Hash, default: {}

property :inet_port, [String, NilClass], default: '0.0.0.0:9001'
property :inet_username, [String, NilClass], default: nil
property :inet_password, [String, NilClass], default: nil

property :include_files, [String, Array], default: lazy {
  "#{supervisord_config_directory}/*.conf"
}

property :template, String, default: 'buyerquest-supervisor'

action :create do
  supervisor_config_directory = new_resource.supervisord_config_directory
  supervisor_config_file = "#{supervisor_config_directory}/supervisord.conf"
  with_run_context :root do
    node.run_state['supervisor'] ||= {}
    node.run_state['supervisor']['directory'] = supervisor_config_directory
    node.run_state['supervisor']['config_file'] = supervisor_config_file
  end
  node.default['supervisor']['inet_http_server']['inet_username'] = new_resource.inet_username
  node.default['supervisor']['inet_http_server']['inet_password'] = new_resource.inet_password
  node.default['supervisor']['inet_http_server']['inet_port'] = new_resource.inet_port

  directory supervisor_config_directory do
    owner 'root'
    group 'root'
    mode '755'
    recursive true
  end

  directory new_resource.supervisord_log_directory do
    owner 'root'
    group 'root'
    mode '755'
    recursive true
  end

  declare_resource(:template, 'supervisord_config_file') do
    cookbook new_resource.template
    path supervisor_config_file
    source 'supervisord.conf.erb'
    owner 'root'
    group 'root'
    mode '644'
    variables config: new_resource
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end


resources/process.rb
# To learn more about Custom Resources, see https://docs.chef.io/custom_resources.html
resource_name :supervisor_process

property :name, String, name_property: true
property :type, String, required: true, equal_to: ['program', 'eventlistener', 'fcgi-program']
# Program
property :command, String, required: true
property :process_name, String, default: '%(program_name)s'
property :numprocs, Integer, default: 1
property :numprocs_start, Integer, default: 0
property :priority, Integer, default: 999
property :autostart, [TrueClass, FalseClass], default: false
property :autorestart, [TrueClass, FalseClass], default: false
property :startsecs, Integer, default: 1
property :startretries, Integer, default: 3
property :exitcodes, Array, default: [0, 2]
property :stopsignal, [String, Symbol], default: :TERM
property :stopwaitsecs, Integer, default: 10
property :stopasgroup,  [TrueClass, FalseClass], default: true
property :killasgroup,  [TrueClass, FalseClass], default: true
property :user, [String, NilClass], default: nil
property :redirect_stderr, [TrueClass, FalseClass], default: false
property :environment, Hash, default: {}
property :directory, [String, NilClass], default: nil
property :umask, [String, NilClass], default: nil
property :serverurl, String, default: 'AUTO'
property :stdout_logfile, String, default: 'AUTO'
property :stdout_logfile_maxbytes, String, default: '50MB'
property :stdout_logfile_backups, Integer, default: 10
property :stdout_capture_maxbytes, String, default: '0'
property :stdout_events_enabled,  [TrueClass, FalseClass], default: false
property :stderr_logfile, String, default: 'AUTO'
property :stderr_logfile_maxbytes, String, default: '50MB'
property :stderr_logfile_backups, Integer, default: 10
property :stderr_capture_maxbytes, String, default: '0'
property :stderr_events_enabled, [TrueClass, FalseClass], default: false

# Eventlistener
property :eventlistener_buffer_size, Integer, default: 10
property :eventlistener_events, Array, default: ['EVENT']
property :eventlistener_result_handler, String, default: 'supervisor.dispatchers:default_handler'

# Fcgi Program
property :fcgi_socket, String, default: 'unix:///var/run/supervisor/%(program_name)s.sock'
property :fcgi_socket_owner, [String, NilClass], default: nil
property :fcgi_socket_mode, String, default: '0700'

property :template, String, default: 'buyerquest-supervisor'

action :create do
  clean_name = new_resource.name.downcase.tr(' ', '_')
  unique_name_for_process = "#{new_resource.type}_#{clean_name}"
  declare_resource(:template, "supervisor_#{unique_name_for_process}") do
    cookbook new_resource.template
    path(lazy { "#{node.run_state['supervisor']['directory']}/#{unique_name_for_process}.conf" })
    source 'process.conf.erb'
    owner 'root'
    group 'root'
    mode '644'
    variables(
      name: clean_name,
      service: new_resource
    )
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end

action :delete do
  clean_name = new_resource.name.downcase.tr(' ', '_')
  unique_name_for_process = "#{new_resource.type}_#{clean_name}"
  file 'delete specific program configuration file' do
    path(lazy { "#{node.run_state['supervisor']['directory']}/#{unique_name_for_process}.conf" })
    action :delete
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end
